client/state/media/reducer.js

```diff
@@ -2,22 +2,156 @@
 import MediaQueryManager from 'lib/query-manager/media';
+import { validateMediaItem } from 'lib/media/utils';
+import { ValidationErrors as MediaValidationErrors } from 'lib/media/constants';
+
+const isExternalMediaError = message =>
+	message.error && ( message.error === 'servicefail' || message.error === 'keyring_token_error' );
+
+const isMediaError = action =>
+	action.error && ( action.siteId || isExternalMediaError( action.error ) );
+
+/**
+ * Returns the updated media errors state after an action has been
+ * dispatched. The state reflects a mapping of site ID, media ID pairing to
+ * an array of errors that occurred for that corresponding media item.
+ *
+ * @param  {object} state  Current state
+ * @param  {object} action Action payload
+ * @returns {object}        Updated state
+ */
+export const errors = ( state = {}, action ) => {
+	switch ( action.type ) {
+		case MEDIA_ITEM_CREATE: {
+			if ( ! action.site || ! action.transientMedia ) {
+				return state;
+			}
+
+			const items = Array.isArray( action.transientMedia )
+				? action.transientMedia
+				: [ action.transientMedia ];
+			const mediaErrors = items.reduce( function( memo, item ) {
+				const itemErrors = validateMediaItem( action.site, item );
+				if ( itemErrors.length ) {
+					memo[ item.ID ] = itemErrors;
+				}
+
+				return memo;
+			}, {} );
+
+			return {
+				...state,
+				[ action.site.ID ]: {
+					...state[ action.site.ID ],
+					...mediaErrors,
+				},
+			};
+		}
+
+		case MEDIA_ITEM_REQUEST_FAILURE:
+		case MEDIA_REQUEST_FAILURE: {
+			// Track any errors which occurred during upload or getting external media
+			if ( ! isMediaError( action ) ) {
+				return state;
+			}
+
+			let mediaErrors;
+
+			if ( Array.isArray( action.error.errors ) ) {
+				mediaErrors = action.error.errors;
+			} else {
+				mediaErrors = [ action.error ];
+			}
+
+			const sanitizedErrors = mediaErrors.map( error => {
+				switch ( error.error ) {
+					case 'http_404':
+						return MediaValidationErrors.UPLOAD_VIA_URL_404;
+					case 'upload_error':
+						if ( error.message.indexOf( 'Not enough space to upload' ) === 0 ) {
+							return MediaValidationErrors.NOT_ENOUGH_SPACE;
+						}
+						if ( error.message.indexOf( 'You have used your space quota' ) === 0 ) {
+							return MediaValidationErrors.EXCEEDS_PLAN_STORAGE_LIMIT;
+						}
+						return MediaValidationErrors.SERVER_ERROR;
+					case 'keyring_token_error':
+						return MediaValidationErrors.SERVICE_AUTH_FAILED;
+					case 'servicefail':
+						return MediaValidationErrors.SERVICE_FAILED;
+					default:
+						return MediaValidationErrors.SERVER_ERROR;
+				}
+			} );
+
+			return {
+				...state,
+				[ action.siteId ]: {
+					...state[ action.siteId ],
+					[ action?.mediaId ?? 0 ]: sanitizedErrors,
+				},
+			};
+		}
+
+		case MEDIA_ERRORS_CLEAR:
+			if ( ! action.siteId ) {
+				return state;
+			}
+
+			return {
+				...omit( state, action.siteId ),
```

Comment:
do you _need_ the omit here? I think the subsequent line overrides it?

Reply:
Good catch, removed in https://github.com/Automattic/wp-calypso/pull/39375/commits/a75fc15ed2117689f9b803ab9b646d0c70b0db63